Allow `paasta rerun` of dependent jobs

macisamuele opened the pull request:
## Pre-PR
paasta rerun will re-execute only the job identified by ``(cluster, service, instance)`` from the command line arguments. It implicitly [removes all the references to the parents](https://github.com/Yelp/paasta/blob/master/paasta_tools/chronos_rerun.py#L145) and [schedules the job to start as soon as possible](https://github.com/Yelp/paasta/blob/master/paasta_tools/chronos_rerun.py#L148).
> 'paasta rerun' creates a copy of the specified PaaSTA scheduled job and
executes it immediately. Parent-dependent relationships are ignored: 'pasta
rerun' only executes individual jobs.

## Post-PR
This PR introduces a new command line argument to the ``rerun`` command. The new argument, ``--rerun-type``, will allow the user to customize the rerun policy used for a specific ``(cluster, service, instance)``:
  * ``--rerun-type instance``: same behaviour of Pre-PR (execute the job stripping out parents and scheduling ASAP)
 * ``--rerun-type graph``: rerun all the dependency graph containing the required job. It will schedule the "scheduled" jobs to start ASAP.

### Example:
<img src="http://i.imgur.com/cumodpH.png"></img>
``paasta rerun -s ... -c ... -i D --rerun-type instance`` will rerun only ``D``
``paasta rerun -s ... -c ... -i D --rerun-type graph`` will rerun **all** the nodes (``A`` will start ASAP)

## Tests
 * Added integration tests on ``chronos_rerun.feature```
 * Unittests for ``chronos_tools``, ``chronos_rerun`` and ``cmd/rerun``
 * skipif OSX tests requiring flock

### NOTE
I have updated ``travis.yaml`` file to let travis run tests, most probably that commit should be reverted before merge!

### FUTURE EXPANSION
To avoid a huge Code Review (since it is already quite big) I'll introduce as a follow-up PR the possibility to run only child jobs.
Using the example of before
``paasta rerun -s ... -c ... -i D --rerun-type childs`` will rerun ``D`` and ``F`` (``D`` will start ASAP)

solarkennedy commented:
This is epic! Give us some time to review it carefully, but thank you!

macisamuele commented:
@solarkennedy no pressure in making a fast CR ... I can understand how delicate the change could be!
In the mean time I see a ``paasta_itest`` failure and tomorrow I'll try to have a look to see if it was a flake or a *real* failure.
As owners of many deeply dependent batches a feature like this could be extremely helpful in rerunning failed jobs. Have a great continuation 😃

solarkennedy commented on paasta_tools/cli/cmds/rerun.py:
```diff
@@ -85,6 +88,21 @@ def add_subparser(subparsers):
         help="define a different soa config directory",
     )
+    rerun_parser.add_argument(
+        '-t', '--rerun-type',
```
What do you think if this was simplified to a single flag called `--rerun-dependent-jobs-too` ?

macisamuele replied:
I was thinking to have a real parameter in order to make this "policy" configurable.
As I wrote in the PR text I would add in a different PR the possibility to run only the required job and all his childs.

solarkennedy replied:
Gotcha, I didn't read carefully enough.

solarkennedy commented on paasta_tools/cli/cmds/rerun.py:
```diff
@@ -85,6 +88,21 @@ def add_subparser(subparsers):
+             "    j1 -> j2, j1 -> j3, j2 -> j3, j2 -> j4\n"
+             "\n"
+             "    Rerunning j2 wih --rerun-type=instance will rerun ONLY j2, j3 and j4 will not be re-ran\n"
+             "    Rerunning j2 wih --rerun-type=graph will rerun j1, j2, j3 and j4\n"
```
Can you make it clear in the help text that it will follow the correct dependent graph ordering?

macisamuele replied:
done

Rob-Johnson commented on paasta_tools/chronos_tools.py:
```diff
@@ -956,3 +961,147 @@ def is_temporary_job(job):
+ENTERED, ENTERED_AND_EXITED = range(0, 2)
+
+
+def dfs(node, get_adjacent_nodes, add_visited, node_status=None, ignore_cycles=False, callback_on_node_enter=None):
```
can we make this function as close to referentially transparent as we can? I had a hard time reasoning about how it worked in ```_get_related_jobs_and_configs```.

macisamuele commented on paasta_itests/steps/setup_steps.py:
```diff
@@ -259,8 +266,9 @@ def call_load_paasta_native_job_config(context):
-@given('we have a deployments.json for the service "{service}" with {disabled} instance "{instance}" image "{image}"')
-def write_soa_dir_deployments(context, service, disabled, instance, image):
+@given('we have a deployments.json for the service "{service}" with {disabled} instance '
+       '"{csv_instancs}" image "{image}"')
```
fixed

macisamuele commented on paasta_itests/steps/chronos_rerun_steps.py:
```diff
@@ -33,6 +33,16 @@ def run_chronos_rerun(context, service_instance):
     context.exit_code, context.output = exit_code, output
 
 
+@when('we run chronos_rerun for service_instance "{service_instance}" with --run-all-related-jobs')
```
nice suggestion ... applied!

macisamuele commented:
I have:
 * rebased the branch on top of the new master (to solve merge conflicts).
 * simplified as much as possible the input parameters of ``dfs`` method (as discussed offline with @Rob-Johnson )
 * Fixed a test failure due to incomplete mocking information

Rob-Johnson commented:
shipit! this looks great to me. I'm a bit concerned about how cleanup_chronos_jobs will affect this, but we can ship and iterate

macisamuele commented:
Hey @Rob-Johnson sorry for the delay in the response.
I have checked the implementation of ``cleanup_chronos_jobs`` and as far as I can get from it we should not have specific issues related to this PR.

``cleanup_chronos_jobs`` tries to clean all the cleanup all the chronos jobs that are currently deployed that are [not "normal" jobs](https://github.com/Yelp/paasta/blob/master/paasta_tools/cleanup_chronos_jobs.py#L154-L155)  (so jobs currently defined in the CONFIGS) and are [not expired temporary jobs](https://github.com/Yelp/paasta/blob/master/paasta_tools/cleanup_chronos_jobs.py#L158).

Having a close look to [``filter_expired_tmp_jobs``](https://github.com/Yelp/paasta/blob/master/paasta_tools/cleanup_chronos_jobs.py#L117) it returns all the jobs that have already [completed their execution](https://github.com/Yelp/paasta/blob/master/paasta_tools/cleanup_chronos_jobs.py#L136).

This logic should not have side effects with the temporary chronos jobs created by this PR, eventually we could try to add a specific test in test_cleanup_chronos_jobs to ensure what I described above.

Let me know on how we would proceed.
